### DISCUSSION TITLE
Default interface implementation that uses references a static abstract interface member

### DISCUSSION THREAD

**Discussion by AlgorithmsAreCool** [dotnet/roslyn#59912] [Q&A]
When I try to provide a default implementation of and inherited interface member that references a static abstract member of my interface, I get a confusing error.

```csharp
public interface IColumnGroupBase<T> : IReadOnlyList<T?>
{
    public static abstract string GetColumnName(int index);
    public static abstract int ColumnCount { get; }

    // CS8926 A static abstract interface member can be accessed only on a type parameter.
    int IReadOnlyCollection<T?>.Count => ColumnCount; 
}

```

1. Is this a supported scenario?
2. I don't understand how you can access something "on a type parameter".

**Comment by RikkiGibson** [comment]
Basically, the only way the runtime can track the necessary type information to call the correct implementation of the static abstract method is when it is called via a constrained type parameter. For example:

```cs
void M<T, U>() where T : IColumnGroupBase<U>
{
    Console.Write(T.ColumnCount);
}
```

**Reply by AlgorithmsAreCool** [reply]
This stuff is mostly above my head, but what I'm hearing is that the type `IColumnGroupBase<T>` cannot be directly used to access static abstract members because we need a type constraint to carry the extra info in the runtime type system. Which is counterintuitive because I would assume the accessing a type directly would be "constrained" to itself. 

So I tried constructing a helper with a constraint, but I get another error when I try to call it using the interface itself as the argument.

```csharp
public interface IColumnGroupBase<T> : IReadOnlyList<T?>
{
    public static abstract string GetColumnName(int index);
    public static abstract int ColumnCount { get; }

    //CS8920: The interface 'IColumnGroupBase<T>' cannot be used as type parameter 'U' in the generic type or method 'IColumnGroupBase<T>.ColumnCountHelper<U>()'. 
    //The constraint interface 'IColumnGroupBase<T>' or its base interface has static abstract members.
    int IReadOnlyCollection<T?>.Count => ColumnCountHelper<IColumnGroupBase<T>>();

    // no error here 🎉
    private static int ColumnCountHelper<U>() where U : IColumnGroupBase<T>
    {
        return U.ColumnCount;
    }
}
```

So the interface itself cannot be used as a argument to the constrained type parameter. When typing this out, I now recall reading something like this [in the design notes.](https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2021/LDM-2021-02-08.md#self-applicability-as-a-constraint).

The LDM designed the feature to disallow an interface containing an abstract static from being able to satisfy a constraint by itself. This applies to inherited interfaces also.

So, I think I understand. When we call an interface member the call is routed to the interface implementation. For a class or struct the compiler requires that we provide an implementation of the member, so there is always one to route to.  This works because at runtime, we have an object instance that carries information in the vtable (sorta...) about which implementation to use.

However, with static abstract interface members, there is no object instance to provide the link to the actual implementation. There might not be any valid implementations loaded, and even if there were, we certainly cannot just pick one out of a hat. Therefore, we can't route the call and there is no way for this to work. So the compiler forbids it.
